zcash_client_sqlite/src/testing/pool.rs

```diff
@@ -318,150 +314,252 @@ pub(crate) fn send_multi_step_proposed_transfer<T: ShieldedPoolTester>() {
+
+        let expected_step0_fee = (zip317::MARGINAL_FEE * 3).unwrap();
+        let expected_step1_fee = zip317::MINIMUM_FEE;
+        let expected_ephemeral = (transfer_amount + expected_step1_fee).unwrap();
+        let expected_step0_change =
+            (value - expected_ephemeral - expected_step0_fee).expect("sufficient funds");
+        assert!(expected_step0_change.is_positive());
+
+        // Generate a ZIP 320 proposal, sending to the wallet's default transparent address
+        // expressed as a TEX address.
+        let tex_addr = match account.usk().default_transparent_address().0 {
+            TransparentAddress::PublicKeyHash(data) => Address::Tex(data),
+            _ => unreachable!(),
+        };
+        let change_memo = Some(Memo::from_str("change").expect("valid memo").encode());
+
+        // We use `st.propose_standard_transfer` here in order to also test round-trip
+        // serialization of the proposal.
+        let proposal = st
+            .propose_standard_transfer::<Infallible>(
+                account.account_id(),
+                StandardFeeRule::Zip317,
+                NonZeroU32::new(1).unwrap(),
+                &tex_addr,
+                transfer_amount,
+                None,
+                change_memo.clone(),
+                T::SHIELDED_PROTOCOL,
+            )
+            .unwrap();
 
-    // Spendable balance matches total balance
-    assert_eq!(st.get_total_balance(account.account_id()), value);
-    assert_eq!(st.get_spendable_balance(account.account_id(), 1), value);
+        let steps: Vec<_> = proposal.steps().iter().cloned().collect();
+        assert_eq!(steps.len(), 2);
+
+        assert_eq!(steps[0].balance().fee_required(), expected_step0_fee);
+        assert_eq!(steps[1].balance().fee_required(), expected_step1_fee);
+        assert_eq!(
+            steps[0].balance().proposed_change(),
+            [
+                ChangeValue::shielded(T::SHIELDED_PROTOCOL, expected_step0_change, change_memo),
+                ChangeValue::ephemeral_transparent((transfer_amount + expected_step1_fee).unwrap()),
+            ]
+        );
+        assert_eq!(steps[1].balance().proposed_change(), []);
 
-    assert_eq!(
-        block_max_scanned(&st.wallet().conn, &st.wallet().params)
-            .unwrap()
-            .unwrap()
-            .block_height(),
-        h
-    );
+        let create_proposed_result = st.create_proposed_transactions::<Infallible, _>(
+            account.usk(),
+            OvkPolicy::Sender,
+            &proposal,
+        );
+        assert_matches!(&create_proposed_result, Ok(txids) if txids.len() == 2);
+        let txids = create_proposed_result.unwrap();
+
+        // Verify that the stored sent outputs match what we're expecting.
+        let mut stmt_sent = st
+            .wallet()
+            .conn
+            .prepare(
+                "SELECT value, to_address, ephemeral_addresses.address, ephemeral_addresses.address_index
```

Comment:
In order for the tests to be migrated to `zcash_client_backend`, this query will have to be performed via a method on `WalletRead` rather than via direct database access. I guess there are a couple of other instances where we use direct database accesses, but it would be really good not to introduce more.

Reply:
Since the previous code also used direct database access, I consider that out-of-scope for this PR.